overlord/snapshotstate/backend/backend.go

```diff
@@ -445,6 +446,169 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
 	return nil
 }
 
+// Import a snapshot from the export file format
+func Import(ctx context.Context, id uint64, r io.Reader) (size int64, snapNames []string, err error) {
+	errPrefix := fmt.Sprintf("cannot import snapshot %d", id)
+
+	// prepare cache location to unpack the import file
+	tempImportDir := path.Join(dirs.SnapCacheDir, "snapshots", fmt.Sprintf("import-%d", id))
+	if err := os.MkdirAll(path.Dir(tempImportDir), 0700); err != nil {
+		return 0, nil, fmt.Errorf("%s: %v", errPrefix, err)
+	}
+	if err := os.Mkdir(tempImportDir, 0700); err != nil {
+		if os.IsExist(err) {
+			return 0, nil, fmt.Errorf("%s: already in progress for this id", errPrefix)
+		}
+		return 0, nil, fmt.Errorf("%s: %v", errPrefix, err)
+	}
+	defer os.RemoveAll(tempImportDir)
+
+	// unpack the tar file to a temporary location and validate.
+	exportFound, size, err := unpackVerifySnapshotImport(r, tempImportDir)
+	if err != nil {
+		return 0, nil, fmt.Errorf("%s: %v", errPrefix, err)
+	}
+	if !exportFound {
+		return 0, nil, fmt.Errorf("%s: no export.json file in uploaded data", errPrefix)
+	}
+
+	// undo all moved snapshots on error
+	var movedSnapshotFiles []string
+	defer func() {
+		if err != nil {
+			for _, fname := range movedSnapshotFiles {
+				if e := os.Remove(fname); e != nil {
+					logger.Noticef("cannot remove imported snapshot file %q: %v", fname, e)
+				}
+			}
+		}
+	}()
+
+	// walk the cache directory to store the snapshot files from the
+	// temp import location to the real location
+	dir, err := osOpen(tempImportDir)
+	if err != nil {
+		return 0, nil, fmt.Errorf("%s: %v", errPrefix, err)
+	}
+	defer dir.Close()
+
+	var readErr error
+	for readErr == nil {
+		var names []string
+		names, readErr = dirNames(dir, 100)
+		if len(names) > 0 {
+			// move the files into place with the new local set ID
+			newSnapNames, newFiles, err := moveCachedSnapshots(tempImportDir, names, id)
```

Comment:
This is not an atomic operation. If we reboot in the middle of this move we will have half an imported snapshot but no indication that the move was incomplete. Also right now on core this move from /var/cache/ to /var/lib gives us an "invalid cross-device link" error which looks like a fun side-effect of our writable-path magic.

Reply:
seems we need to discuss what to do about these points

Reply:
I chatted with Samuele and we will fix this by having a "$sid_importing" file that will be created first. The backend list code will skip all files with an importing setID. When the improt is finished this file is removed again.

Reply:
The commit https://github.com/snapcore/snapd/pull/9036/commits/7f833c2c5b5d751be2b67204e4166e4e891553e7 has a sketch of my proposal to solve the issue, namely:
1. the Iter() code ignores any setID that has a $setID_importing filename
2. when we start importing we write *first* a $setID_importing filename
3. we import into the regular snapshot dir, this avoid all cross-device moves or non-atomic multiple-files moves
4. we either Commit() or Cancel(), i.e. either remove the $setID_importing filename (commit) or do `rm $setID_*.zip` to cancel.
5. on startup we find all leftover $setID_imprting files and cleanup the "$setID_*.zip"

Reply:
@mvo5 consider just using file locks, they seem like an easier solution as this removes 5. (on startup they are not locked), extra stat calls in 2, etc.

Reply:
@zyga happy to chat about this but (5) we will always need, what if we restart in the middle of an import? we always need to cleanup

Fwiw, I pushed a full version (minus tests and cleanup on startup) in https://github.com/mvo5/snappy/tree/snapshot-import-3

Reply:
I misread point 5., please disregard my comment on that.